File: project/.sdk/tm/go/test/custom_utility_test.go
package sdktest

import (
	"testing"

	sdk "GOMODULE"
)

func TestCustomUtility(t *testing.T) {
	t.Run("basic", func(t *testing.T) {
		client := sdk.TestSDK(nil, map[string]any{
			"apikey": "APIKEY01",
			"utility": map[string]any{
				"auth":       func() map[string]any { return map[string]any{"util": "AUTH"} },
				"body":       func() map[string]any { return map[string]any{"util": "BODY"} },
				"contextify": func() map[string]any { return map[string]any{"util": "CONTEXTIFY"} },
				"done":       func() map[string]any { return map[string]any{"util": "DONE"} },
				"error":      func() map[string]any { return map[string]any{"util": "ERROR"} },
				"findparam":  func() map[string]any { return map[string]any{"util": "FINDPARAM"} },
				"fullurl":    func() map[string]any { return map[string]any{"util": "FULLURL"} },
				"headers":    func() map[string]any { return map[string]any{"util": "HEADERS"} },
				"method":     func() map[string]any { return map[string]any{"util": "METHOD"} },
				"operator":   func() map[string]any { return map[string]any{"util": "OPERATOR"} },
				"params":     func() map[string]any { return map[string]any{"util": "PARAMS"} },
				"query":      func() map[string]any { return map[string]any{"util": "QUERY"} },
				"reqform":    func() map[string]any { return map[string]any{"util": "REQFORM"} },
				"request":    func() map[string]any { return map[string]any{"util": "REQUEST"} },
				"resbasic":   func() map[string]any { return map[string]any{"util": "RESBASIC"} },
				"resbody":    func() map[string]any { return map[string]any{"util": "RESBODY"} },
				"resform":    func() map[string]any { return map[string]any{"util": "RESFORM"} },
				"resheaders": func() map[string]any { return map[string]any{"util": "RESHEADERS"} },
				"response":   func() map[string]any { return map[string]any{"util": "RESPONSE"} },
				"result":     func() map[string]any { return map[string]any{"util": "RESULT"} },
				"spec":       func() map[string]any { return map[string]any{"util": "SPEC"} },
			},
		})

		u := client.GetUtility()

		checks := map[string]string{
			"auth":       "AUTH",
			"body":       "BODY",
			"contextify": "CONTEXTIFY",
			"done":       "DONE",
			"error":      "ERROR",
			"findparam":  "FINDPARAM",
			"fullurl":    "FULLURL",
			"headers":    "HEADERS",
			"method":     "METHOD",
			"operator":   "OPERATOR",
			"params":     "PARAMS",
			"query":      "QUERY",
			"reqform":    "REQFORM",
			"request":    "REQUEST",
			"resbasic":   "RESBASIC",
			"resbody":    "RESBODY",
			"resform":    "RESFORM",
			"resheaders": "RESHEADERS",
			"response":   "RESPONSE",
			"result":     "RESULT",
			"spec":       "SPEC",
		}

		for key, expected := range checks {
			fn, ok := u.Custom[key]
			if !ok {
				t.Errorf("expected custom utility %q to exist", key)
				continue
			}
			if f, ok := fn.(func() map[string]any); ok {
				result := f()
				if result["util"] != expected {
					t.Errorf("custom utility %q: got %v, want %v", key, result["util"], expected)
				}
			} else {
				t.Errorf("custom utility %q: expected func() map[string]any, got %T", key, fn)
			}
		}
	})
}

File: ts/project/.sdk/tm/cpp/test/custom_utility_test.cpp
// ProjectName SDK — custom utility overrides supplied via options.utility
// land on the utility object's custom map (mirrors java
// test/CustomUtilityTest.java). This exercises value-semantics gotcha #8:
// the struct clone in makeOptions must NOT drop the function (Injector)
// values, so each override is still callable after client construction.

#include "harness.hpp"

using namespace sdk;
using namespace sdk::fh;

// A callable override: an Injector that ignores its args and returns { util: tag }.
static Value mkutil(const std::string& tag) {
  vs::Injector fn = [tag](vs::Injection&, const Value&, const std::string&, const Value&) -> Value {
    return vmap({{"util", Value(tag)}});
  };
  return Value(fn);
}

static std::string upper(const std::string& s) {
  std::string o = s;
  for (auto& c : o) c = static_cast<char>(std::toupper((unsigned char)c));
  return o;
}

static void basic() {
  const std::vector<std::string> keys = {
      "auth", "body", "contextify", "done", "error", "findparam", "fullurl",
      "headers", "method", "operator", "params", "query", "reqform",
      "request", "resbasic", "resbody", "resform", "resheaders", "response",
      "result", "spec",
  };

  Value customUtils = vmap();
  for (const auto& k : keys) map_put(customUtils, k, mkutil(upper(k)));

  auto client = ProjectNameSDK::testSDK(
      Value::undef(),
      fhMap({{"apikey", Value("APIKEY01")}, {"utility", customUtils}}));

  UtilityPtr u = client->getUtility();

  for (const auto& k : keys) {
    Value fn = getp(u->custom, k);
    ASSERT_TRUE(fn.is_injector(), "expected custom utility " + k + " to exist");
    if (fn.is_injector()) {
      vs::Injection inj(Value::undef(), Value::undef());
      Value r = fn.as_injector()(inj, Value::undef(), std::string(""), Value::undef());
      ASSERT_EQ_VAL(getp(r, "util"), Value(upper(k)), "custom utility " + k);
    }
  }
}

int main() {
  T_RUN(basic);
  return sdktest::summary("custom_utility_test");
}
